nano-zkp/src/gadgets/mnt4/hash_to_curve.rs

```diff
@@ -78,69 +78,112 @@ impl HashToCurve {
-            x_bits.iter().map(|i| i.value().unwrap()).collect(),
+        let (nonce_bits, hash_point) = Self::try_and_increment(
+            c0_bits.iter().map(|i| i.value().unwrap()).collect(),
+            c1_bits.iter().map(|i| i.value().unwrap()).collect(),
+            c2_bits.iter().map(|i| i.value().unwrap()).collect(),
             y_bit.value()?,
         );
 
-        // Allocate the nonce bits and convert to a field element.
+        // Allocate the nonce bits.
         let nonce_bits_var = Vec::<Boolean<MNT4Fr>>::new_witness(cs.clone(), || Ok(nonce_bits))?;
-        let nonce = Boolean::le_bits_to_fp_var(&nonce_bits_var)?;
 
-        // Allocate the G1 hash point.
-        let g1_var = G1Var::new_witness(cs.clone(), || Ok(g1))?;
+        // Allocate the hash point.
+        let hash_var = G2Var::new_witness(cs.clone(), || Ok(hash_point))?;
 
         // Convert the x-coordinate bits into a field element.
-        x_bits.reverse();
-        let x = Boolean::le_bits_to_fp_var(&x_bits)?;
+        c0_bits.reverse();
+        let c0 = Boolean::le_bits_to_fp_var(&c0_bits)?;
+
+        c1_bits.reverse();
+        let c1 = Boolean::le_bits_to_fp_var(&c1_bits)?;
+
+        c2_bits.reverse();
+        let c2 = Boolean::le_bits_to_fp_var(&c2_bits)?;
+
+        let x = Fq3Var::new(c0, c1, c2);
+
+        // Convert the nonce bits into a field element. Using double-and-add.
+        let mut nonce = Fq3Var::zero();
+        let mut power = Fq3Var::one();
+
+        for bit in nonce_bits_var {
+            let new_nonce = &nonce + &power;
+
+            nonce = Fq3Var::conditionally_select(&bit, &new_nonce, &nonce)?;
+
+            power = power.double()?;
+        }
 
         // Add the nonce to the x-coordinate.
         let x = x + nonce;
 
-        // Compare the coordinates of our G1 hash point to the calculated coordinates.
-        let g1_var_affine = g1_var.to_affine()?;
+        // Compare the coordinates of our hash point to the calculated coordinates.
+        let hash_var_affine = hash_var.to_affine()?;
 
-        let y_coordinate = YToBitGadget::y_to_bit_g1(cs, &g1_var_affine)?;
-        let coordinates = g1_var_affine.to_constraint_field()?;
+        let y_coordinate = YToBitGadget::y_to_bit_g2(cs, &hash_var_affine)?;
+        let x_coordinate = hash_var_affine.x;
+        let inf_coordinate = hash_var_affine.infinity;
 
-        coordinates[0].enforce_equal(&x)?;
+        x_coordinate.enforce_equal(&x)?;
         y_coordinate.enforce_equal(y_bit)?;
-        coordinates[2].enforce_equal(&FpVar::zero())?;
+        inf_coordinate.enforce_equal(&Boolean::constant(false))?;
 
-        // We don't need to scale by the cofactor since MNT6-753 has a cofactor of one.
+        // TODO! The hash point still needs to be scaled by the cofactor of the curve. Otherwise the
+        //  signature won't verify.
+        // We now scale by the cofactor.
+        //let scaled_hash_var = hash_var_affine.fixed_scalar_mul_le()?;
```

Comment:
Honestly, because it is more complicated than it seems and is not really needed now. I prefer to merge this PR first and then I can implement this part when I'm doing the Poseidon hash PR.